### PR TITLE
Features/panic clean

### PR DESCRIPTION
**Pull request by bird512**

### CODE REVIEW

File: common/privatekey.go [hunk 1]
```diff
@@ -47,14 +46,14 @@ func (pk PrivateKey) Sign(hash []byte) Sign {
 	if err == nil {
 		sign = *BytesToSign(sig)
 	} else {
-		panic(fmt.Sprintf("Sign Failed, reason : %v.\n", err.Error()))
+		err = DefaultLogger.Errorf("Sign Failed, reason : %v.\n", err.Error())
```
Comment (darren0718) [inline common/privatekey.go]: should use fmt.Errorf. DefaultLogger maybe unavailable in test case

File: common/privatekey.go [hunk 2]
```diff
@@ -66,9 +65,10 @@ func GenerateKey(s string) PrivateKey {
 	if err == nil {
 		pk.PrivKey = *_pk
 	} else {
-		panic(fmt.Sprintf("GenKey Failed, reason : %v.\n", err.Error()))
+		err = DefaultLogger.Errorf("GenKey Failed, reason : %v.\n", err.Error())
```
Comment (darren0718) [inline common/privatekey.go]: same as above

File: consensus/logical/group_info.go [hunk 3]
```diff
@@ -99,7 +100,7 @@ func (sgi StaticGroupInfo) GetMinerPos(id groupsig.ID) int {
 		pos = v
 		// Double verification
 		if !sgi.GInfo.Mems[pos].IsEqual(id) {
-			panic("double check fail!id=" + id.GetHexString())
+			pos = -1
```
Comment (darren0718) [inline consensus/logical/group_info.go]: should panic. index different from mem-array position, somothing error

File: consensus/logical/group_info.go [hunk 4]
```diff
@@ -307,8 +308,11 @@ func (gg *GlobalGroups) getGroupFromCache(id groupsig.ID) (g *StaticGroupInfo, e
 	index, ok := gg.gIndex[id.GetHexString()]
 	if ok {
 		g, err = gg.getGroupByIndex(index)
+		if err != nil {
+			return
+		}
 		if !g.GroupID.IsEqual(id) {
-			panic("ggIndex error")
+			err = errors.New("ggIndex error")
```
Comment (darren0718) [inline consensus/logical/group_info.go]: should panic

File: consensus/logical/group_info.go [hunk 5]
```diff
@@ -487,13 +491,13 @@ func (gg *GlobalGroups) removeGroups(gids []groupsig.ID) {
...
-func (gg *GlobalGroups) getGenesisGroup() *StaticGroupInfo {
+func (gg *GlobalGroups) getGenesisGroup() (*StaticGroupInfo, error) {
 	if gg.GetGroupSize() == 0 {
-		return nil
+		return nil, nil
 	}
 	g := gg.groups[0]
 	if g.GInfo.GI.GHeader.WorkHeight != 0 {
-		panic("genesis group error")
+		return g, errors.New("genesis group error")
```
Comment (darren0718) [inline consensus/logical/group_info.go]: should panic

File: consensus/logical/message_processor_group.go [hunk 6]
```diff
@@ -218,6 +222,7 @@ func (p *Processor) OnMessageGroupInit(msg *model.ConsensusGroupRawMessage) {
 	tlog := newHashTraceLog("OMGI", gHash, msg.SI.GetID())
 
 	if msg.SI.DataHash != msg.GenHash() || gh.Hash != gh.GenHash() {
+		// hold it for now
 		panic("msg gis hash diff")
```
Comment (darren0718) [inline consensus/logical/message_processor_group.go]: return directly replace of panic for illegal user input

**Review by darren0718 — CHANGES_REQUESTED** [review]

File: consensus/logical/message_processor_group.go (hunk 6 above)
Comment (bird512) [reply consensus/logical/message_processor_group.go]: done

File: common/privatekey.go (hunk 1 above)
Comment (bird512) [reply common/privatekey.go]: All test cases run success

File: common/privatekey.go (hunk 2 above)
Comment (bird512) [reply common/privatekey.go]: same as above

File: consensus/logical/group_info.go (hunk 3 above)
Comment (bird512) [reply consensus/logical/group_info.go]: done

File: consensus/logical/group_info.go (hunk 4 above)
Comment (bird512) [reply consensus/logical/group_info.go]: done

File: consensus/logical/group_info.go (hunk 5 above)
Comment (bird512) [reply consensus/logical/group_info.go]: done

File: common/privatekey.go (hunk 1 above)
Comment (bird512) [reply common/privatekey.go]: done

File: common/privatekey.go (hunk 2 above)
Comment (bird512) [reply common/privatekey.go]: done

**Review by darren0718 — APPROVED** [review]

**Pull request merged**